Documentation/ceph-best-practices.md

```diff
@@ -0,0 +1,516 @@
+use cases. To limit Ceph to specific nodes, we can label Kubernetes nodes and configure Rook to have
+Affinity (as a hard preference) for our label.
+
+For our example, let's label the desired storage nodes with `storage-node=true`. To run Rook and
+Ceph daemons on labeled nodes, we will configure Affinities for this label in both the Rook
+Operator manifest (`operator.yaml`) and the Ceph cluster manifest (`cluster.yaml`).
+
+For Rook daemons and the CSI driver daemons, adjust the Operator manifest. Note that the CSI
+Provisioner is best started on the same nodes as other Ceph daemons.
+
+**`operator.yaml`:**
+```yaml
+CSI_PROVISIONER_NODE_AFFINITY: “storage-node=true”
+AGENT_NODE_AFFINITY: “storage-node=true”
+DISCOVER_AGENT_NODE_AFFINITY: “storage-node=true”
+```
+
+For `all` Ceph daemons, we must add Affinity for storage nodes in `cluster.yaml`.
+
+**`cluster.yaml`:**
+```yaml
+placement:
+  all:
+    nodeAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+        nodeSelectorTerms:
+          - matchExpressions:
+              - key: storage-node
+                operator: In
+                values:
+                  - "true"
+```
+
+## Segregating Ceph from User Applications
+
+You could also have reason to totally separate Rook and Ceph nodes from application nodes. This
+falls under the disaggregated use-case, and it is a more traditional way to deploy storage. In this
+case, we still need to follow the best practices described in the section above to
+[Limit Ceph to Specific Nodes](#limiting-ceph-to-specific-nodes), and we also need some additional
+settings.
+
+To segregate Ceph from user applications, we will also label all non-storage nodes with
+`storage-node=false`. The CSI plugin pods must run where user application Pods run and not where
+Rook or Ceph Pods are run. Add a CSI plugin Affinity for all non-storage nodes in the Rook
+operator configuration.
+
+**`operator.yaml`:**
+```yaml
+CSI_PLUGIN_NODE_AFFINITY: "storage-node=false"
+```
+
+In addition to that, we will set Kubernetes node Taints and configure Rook Tolerations. For example,
+Taint the storage nodes with `storage-node=true:NoSchedule`, and then add the Tolerations shown
+below to the Rook operator in `operator.yaml`.
+
+**`operator.yaml`:**
+```yaml
+AGENT_TOLERATIONS: |
+  - key: storage-node
+    operator: Exists
+```
+```yaml
+DISCOVER_TOLERATIONS: |
+  - key: storage-node
+    operator: Exists
+```
+```yaml
+CSI_PROVISIONER_TOLERATIONS: |
+  - key: storage-node
+    operator: Exists
+```
+
+We must also add a Toleration for `all` Ceph daemon Pods in `cluster.yaml`.
+
+**`cluster.yaml`:**
+```yaml
+placement:
+  all:
+    tolerations:
+      - key: storage-node
+        operator: Exists
+```
+
+## Setting Ceph CRUSH Map via Kubernetes Node Labels
+
+A feature that was implemented early in Rook’s development is to set Ceph’s CRUSH map via Kubernetes
+Node labels. For our example data center, we recommend labelling Nodes with `room`, `rack`, and
```

Comment:
Suggestion for this paragraph:
```
For Ceph to optimally handle failure domains, it is recommended to label the nodes so that Rook can
configure the OSDs in a CRUSH hierarchy that will match your datacenter failure domain hierarchy. 
If no topology labels are added to the nodes, the only failure domain considered by the cluster will 
be at the host layer. If you have another level of outage, Ceph won't be able to handle that scenario 
as gracefully. 

Note that every location in the hierarchy must be uniquely named. You cannot have two failure domains with the same name at any level.

In the example data center example above, labeling nodes will look like the following:
```